python/lsst/verify/ap/dataset.py

```diff
@@ -0,0 +1,218 @@
+    dataset, barring concurrent changes to the file system or EUPS operations.
+    Constructing a Dataset does not create a compatible output repository(ies),
+    which can be done by calling `makeOutputRepo`.
+
+    Parameters
+    ----------
+    dataset_id : `str`
+       A tag identifying the dataset.
+
+    Raises
+    ------
+    `RuntimeError`:
+        `dataset_id` exists, but is not correctly organized or incomplete
+    `ValueError`:
+        `dataset_id` is not a recognized dataset. No side effects if this
+        exception is raised.
+    """
+
+    def __init__(self, dataset_id):
+        try:
+            dataset_package = self._getDatasetInfo()[dataset_id]
+        except KeyError:
+            raise ValueError('Unsupported dataset: ' + dataset_id)
+
+        self._data_root_dir = getPackageDir(dataset_package)
+        self._validate_package()
+
+        self._init_package(dataset_package)
+
+    def _init_package(self, name):
+        """Load the package backing this dataset.
+
+        Parameters
+        ----------
+        name : `str`
+           The EUPS package identifier for the desired package.
+        """
+        Eups().setup(name)
+
+    @staticmethod
+    def get_supported_datasets():
+        """The dataset IDs that can be passed to this class's constructor.
+
+        Returns
+        -------
+        A set of strings of valid tags
+
+        Raises
+        ------
+        `IoError`:
+            if the config file does not exist or is not readable
+        `RuntimeError`:
+            if the config file exists, but does not contain the expected data
+        """
+        return Dataset._getDatasetInfo().keys()
+
+    @staticmethod
+    def _getDatasetInfo():
+        """Return external data on supported datasets.
+
+        If an exception is raised, the program state shall be unchanged.
+
+        Returns
+        -------
+        A map from dataset IDs to package names.
+
+        Raises
+        ------
+        `IoError`:
+            the config file does not exist or is not readable
+        `RuntimeError`:
+            the config file exists, but does not contain the expected data
+        """
+        if not hasattr(Dataset, '_dataset_config'):
+            try:
+                yaml_file = os.path.join(getPackageDir('verify_ap'), 'config/dataset_config.yaml')
+                with open(yaml_file, 'r') as config:
+                    dataset_map = yaml.safe_load(config)['datasets']
+                if not isinstance(dataset_map, dict):
+                    raise TypeError('`datasets` is not a dictionary')
+
+                Dataset._dataset_config = dataset_map
+            except (KeyError, TypeError, yaml.YAMLError) as e:
+                raise_from(RuntimeError('Invalid config file.'), e)
+
+        return Dataset._dataset_config
+
+    @property
+    def dataset_root(self):
+        """The parent directory containing everything related to the dataset.
```

Comment:
I think the code doesn't make it immediately clear that `dataset_root` will know where to find "lower-level" things like `data_location`, `calib_location`, etc. But I guess there's probably not another obvious way to define them all, so it's fine.